File: cpp_bind/character.cpp
#include "character.h"
#include "BulletDynamics/Character/btKinematicCharacterController.h"
#include "BulletCollision/CollisionDispatch/btGhostObject.h"
#include "BulletCollision/CollisionShapes/btConvexShape.h"

nkCharacterHandle nk_character_create(nkGhostObjectHandle ghost, nkShapeHandle convex_shape, 
    nkReal step_height, nkReal up_x, nkReal up_y, nkReal up_z)
{
    if (!ghost || !convex_shape) return nullptr;
    
    btPairCachingGhostObject* ghost_object = static_cast<btPairCachingGhostObject*>(ghost);
    btConvexShape* convex = static_cast<btConvexShape*>(convex_shape);
    btVector3 up(up_x, up_y, up_z);
    
    btKinematicCharacterController* character = new btKinematicCharacterController(ghost_object, convex, step_height, up);
    return static_cast<nkCharacterHandle>(character);
}

void nk_character_destroy(nkCharacterHandle character)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    delete controller;
}

void nk_character_set_walk_direction(nkCharacterHandle character, nkReal x, nkReal y, nkReal z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setWalkDirection(btVector3(x, y, z));
}

void nk_character_set_velocity_for_time(nkCharacterHandle character, nkReal vx, nkReal vy, nkReal vz, nkReal time_interval)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setVelocityForTimeInterval(btVector3(vx, vy, vz), time_interval);
}

void nk_character_set_linear_velocity(nkCharacterHandle character, nkReal x, nkReal y, nkReal z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setLinearVelocity(btVector3(x, y, z));
}

void nk_character_get_linear_velocity(nkCharacterHandle character, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    btVector3 velocity = controller->getLinearVelocity();
    *out_x = velocity.x();
    *out_y = velocity.y();
    *out_z = velocity.z();
}

void nk_character_set_angular_velocity(nkCharacterHandle character, nkReal x, nkReal y, nkReal z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setAngularVelocity(btVector3(x, y, z));
}

void nk_character_get_angular_velocity(nkCharacterHandle character, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    btVector3 velocity = controller->getAngularVelocity();
    *out_x = velocity.x();
    *out_y = velocity.y();
    *out_z = velocity.z();
}

void nk_character_jump(nkCharacterHandle character, nkReal vx, nkReal vy, nkReal vz)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->jump(btVector3(vx, vy, vz));
}

int nk_character_can_jump(nkCharacterHandle character)
{
    if (!character) return 0;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->canJump() ? 1 : 0;
}

int nk_character_on_ground(nkCharacterHandle character)
{
    if (!character) return 0;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->onGround() ? 1 : 0;
}

void nk_character_set_jump_speed(nkCharacterHandle character, nkReal speed)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setJumpSpeed(speed);
}

nkReal nk_character_get_jump_speed(nkCharacterHandle character)
{
    if (!character) return 0.0f;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->getJumpSpeed();
}

void nk_character_set_fall_speed(nkCharacterHandle character, nkReal speed)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setFallSpeed(speed);
}

nkReal nk_character_get_fall_speed(nkCharacterHandle character)
{
    if (!character) return 0.0f;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->getFallSpeed();
}

void nk_character_set_max_jump_height(nkCharacterHandle character, nkReal height)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setMaxJumpHeight(height);
}

void nk_character_set_max_slope(nkCharacterHandle character, nkReal slope_radians)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setMaxSlope(slope_radians);
}

nkReal nk_character_get_max_slope(nkCharacterHandle character)
{
    if (!character) return 0.0f;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->getMaxSlope();
}

void nk_character_set_step_height(nkCharacterHandle character, nkReal height)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setStepHeight(height);
}

nkReal nk_character_get_step_height(nkCharacterHandle character)
{
    if (!character) return 0.0f;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->getStepHeight();
}

void nk_character_set_gravity(nkCharacterHandle character, nkReal x, nkReal y, nkReal z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setGravity(btVector3(x, y, z));
}

void nk_character_get_gravity(nkCharacterHandle character, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    btVector3 gravity = controller->getGravity();
    *out_x = gravity.x();
    *out_y = gravity.y();
    *out_z = gravity.z();
}

void nk_character_set_up(nkCharacterHandle character, nkReal x, nkReal y, nkReal z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setUp(btVector3(x, y, z));
}

void nk_character_get_up(nkCharacterHandle character, nkReal* out_x, nkReal* out_y, nkReal* out_z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    btVector3 up = controller->getUp();
    *out_x = up.x();
    *out_y = up.y();
    *out_z = up.z();
}

void nk_character_warp(nkCharacterHandle character, nkReal x, nkReal y, nkReal z)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->warp(btVector3(x, y, z));
}

void nk_character_reset(nkCharacterHandle character)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->reset(nullptr);
}

void nk_character_set_linear_damping(nkCharacterHandle character, nkReal damping)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setLinearDamping(damping);
}

nkReal nk_character_get_linear_damping(nkCharacterHandle character)
{
    if (!character) return 0.0f;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->getLinearDamping();
}

void nk_character_set_angular_damping(nkCharacterHandle character, nkReal damping)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setAngularDamping(damping);
}

nkReal nk_character_get_angular_damping(nkCharacterHandle character)
{
    if (!character) return 0.0f;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->getAngularDamping();
}

void nk_character_set_max_penetration_depth(nkCharacterHandle character, nkReal depth)
{
    if (!character) return;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    controller->setMaxPenetrationDepth(depth);
}

nkReal nk_character_get_max_penetration_depth(nkCharacterHandle character)
{
    if (!character) return 0.0f;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return controller->getMaxPenetrationDepth();
}

nkGhostObjectHandle nk_character_get_ghost_object(nkCharacterHandle character)
{
    if (!character) return nullptr;
    btKinematicCharacterController* controller = static_cast<btKinematicCharacterController*>(character);
    return static_cast<nkGhostObjectHandle>(controller->getGhostObject());
}

File: rust/src/core/character.rs
use std::ffi::c_void;
use std::ptr::NonNull;

use super::types::{Real, Vec3};
use super::ghost::GhostObject;
use super::collision::ShapeHandle;
use crate::ffi;

pub struct CharacterController {
    handle: NonNull<c_void>,
}

impl CharacterController {
    pub fn new(ghost: &GhostObject, convex_shape: &ShapeHandle, step_height: Real, up: Vec3) -> Self {
        let handle = unsafe {
            ffi::nk_character_create(ghost.handle(), convex_shape.handle(), step_height, up.x, up.y, up.z)
        };
        Self {
            handle: NonNull::new(handle).expect("Failed to create character controller"),
        }
    }

    pub fn handle(&self) -> *mut c_void {
        self.handle.as_ptr()
    }

    pub fn set_walk_direction(&self, direction: Vec3) {
        unsafe {
            ffi::nk_character_set_walk_direction(self.handle.as_ptr(), direction.x, direction.y, direction.z);
        }
    }

    pub fn set_velocity_for_time(&self, velocity: Vec3, time_interval: Real) {
        unsafe {
            ffi::nk_character_set_velocity_for_time(
                self.handle.as_ptr(),
                velocity.x, velocity.y, velocity.z,
                time_interval,
            );
        }
    }

    pub fn set_linear_velocity(&self, velocity: Vec3) {
        unsafe {
            ffi::nk_character_set_linear_velocity(self.handle.as_ptr(), velocity.x, velocity.y, velocity.z);
        }
    }

    pub fn linear_velocity(&self) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_character_get_linear_velocity(self.handle.as_ptr(), &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn set_angular_velocity(&self, velocity: Vec3) {
        unsafe {
            ffi::nk_character_set_angular_velocity(self.handle.as_ptr(), velocity.x, velocity.y, velocity.z);
        }
    }

    pub fn angular_velocity(&self) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_character_get_angular_velocity(self.handle.as_ptr(), &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn jump(&self, velocity: Vec3) {
        unsafe {
            ffi::nk_character_jump(self.handle.as_ptr(), velocity.x, velocity.y, velocity.z);
        }
    }

    pub fn can_jump(&self) -> bool {
        unsafe { ffi::nk_character_can_jump(self.handle.as_ptr()) != 0 }
    }

    pub fn on_ground(&self) -> bool {
        unsafe { ffi::nk_character_on_ground(self.handle.as_ptr()) != 0 }
    }

    pub fn set_jump_speed(&self, speed: Real) {
        unsafe {
            ffi::nk_character_set_jump_speed(self.handle.as_ptr(), speed);
        }
    }

    pub fn jump_speed(&self) -> Real {
        unsafe { ffi::nk_character_get_jump_speed(self.handle.as_ptr()) }
    }

    pub fn set_fall_speed(&self, speed: Real) {
        unsafe {
            ffi::nk_character_set_fall_speed(self.handle.as_ptr(), speed);
        }
    }

    pub fn fall_speed(&self) -> Real {
        unsafe { ffi::nk_character_get_fall_speed(self.handle.as_ptr()) }
    }

    pub fn set_max_jump_height(&self, height: Real) {
        unsafe {
            ffi::nk_character_set_max_jump_height(self.handle.as_ptr(), height);
        }
    }

    pub fn set_max_slope(&self, slope_radians: Real) {
        unsafe {
            ffi::nk_character_set_max_slope(self.handle.as_ptr(), slope_radians);
        }
    }

    pub fn max_slope(&self) -> Real {
        unsafe { ffi::nk_character_get_max_slope(self.handle.as_ptr()) }
    }

    pub fn set_step_height(&self, height: Real) {
        unsafe {
            ffi::nk_character_set_step_height(self.handle.as_ptr(), height);
        }
    }

    pub fn step_height(&self) -> Real {
        unsafe { ffi::nk_character_get_step_height(self.handle.as_ptr()) }
    }

    pub fn set_gravity(&self, gravity: Vec3) {
        unsafe {
            ffi::nk_character_set_gravity(self.handle.as_ptr(), gravity.x, gravity.y, gravity.z);
        }
    }

    pub fn gravity(&self) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_character_get_gravity(self.handle.as_ptr(), &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn set_up(&self, up: Vec3) {
        unsafe {
            ffi::nk_character_set_up(self.handle.as_ptr(), up.x, up.y, up.z);
        }
    }

    pub fn up(&self) -> Vec3 {
        let mut x: Real = 0.0;
        let mut y: Real = 0.0;
        let mut z: Real = 0.0;
        unsafe {
            ffi::nk_character_get_up(self.handle.as_ptr(), &mut x, &mut y, &mut z);
        }
        Vec3::new(x, y, z)
    }

    pub fn warp(&self, position: Vec3) {
        unsafe {
            ffi::nk_character_warp(self.handle.as_ptr(), position.x, position.y, position.z);
        }
    }

    pub fn reset(&self) {
        unsafe {
            ffi::nk_character_reset(self.handle.as_ptr());
        }
    }

    pub fn set_linear_damping(&self, damping: Real) {
        unsafe {
            ffi::nk_character_set_linear_damping(self.handle.as_ptr(), damping);
        }
    }

    pub fn linear_damping(&self) -> Real {
        unsafe { ffi::nk_character_get_linear_damping(self.handle.as_ptr()) }
    }

    pub fn set_angular_damping(&self, damping: Real) {
        unsafe {
            ffi::nk_character_set_angular_damping(self.handle.as_ptr(), damping);
        }
    }

    pub fn angular_damping(&self) -> Real {
        unsafe { ffi::nk_character_get_angular_damping(self.handle.as_ptr()) }
    }

    pub fn set_max_penetration_depth(&self, depth: Real) {
        unsafe {
            ffi::nk_character_set_max_penetration_depth(self.handle.as_ptr(), depth);
        }
    }

    pub fn max_penetration_depth(&self) -> Real {
        unsafe { ffi::nk_character_get_max_penetration_depth(self.handle.as_ptr()) }
    }

    pub fn ghost_object(&self) -> *mut c_void {
        unsafe { ffi::nk_character_get_ghost_object(self.handle.as_ptr()) }
    }
}

impl Drop for CharacterController {
    fn drop(&mut self) {
        unsafe {
            ffi::nk_character_destroy(self.handle.as_ptr());
        }
    }
}

unsafe impl Send for CharacterController {}
unsafe impl Sync for CharacterController {}
